airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py

```diff
@@ -0,0 +1,252 @@
+
+    def get(self, key: str, default: Any) -> Any:
+        return self._stream_slice.get(key, default)
+
+    def __eq__(self, other):
+        if type(other) == dict:
+            return self._stream_slice == other
+        if type(other) == PerPartitionStreamSlice:
+            # noinspection PyProtectedMember
+            return self._partition == other._partition and self._cursor_slice == other._cursor_slice
+        return False
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+
+class PerPartitionCursor(StreamSlicer):
+    """
+    Given a stream has many partitions, it is important to provide a state per partition.
+
+    Record | Stream Slice | Last Record | DatetimeCursorBased cursor
+    -- | -- | -- | --
+    1 | {"start_time": "2021-01-01","end_time": "2021-01-31","owner_resource": "1"''} | cursor_field: “2021-01-15” | 2021-01-15
+    2 | {"start_time": "2021-02-01","end_time": "2021-02-28","owner_resource": "1"''} | cursor_field: “2021-02-15” | 2021-02-15
+    3 | {"start_time": "2021-01-01","end_time": "2021-01-31","owner_resource": "2"''} | cursor_field: “2021-01-03” | 2021-01-03
+    4 | {"start_time": "2021-02-01","end_time": "2021-02-28","owner_resource": "2"''} | cursor_field: “2021-02-14” | 2021-02-14
+
+    Given the following errors, this can lead to some loss or duplication of records:
+    When | Problem | Affected Record
+    -- | -- | --
+    Between record #1 and #2 | Loss | #3
+    Between record #2 and #3 | Loss | #3, #4
+    Between record #3 and #4 | Duplication | #1, #2
+
+    Therefore, we need to manage state per partition.
+    """
+
+    _NO_STATE = {}
+    _NO_CURSOR_STATE = {}
+    _KEY = 0
+    _VALUE = 1
+
+    def __init__(self, cursor_factory, partition_router):
+        self._cursor_factory = cursor_factory
+        self._partition_router = partition_router
+        self._cursor_per_partition = {}
+
+    def stream_slices(self, sync_mode: SyncMode, stream_state: StreamState) -> Iterable[PerPartitionStreamSlice]:
+        """
+        We knowingly avoid using stream_state as we want PerPartitionCursor to manage its own state.
+        """
+        slices = self._partition_router.stream_slices(sync_mode, self._NO_STATE)
+        for partition in slices:
+            cursor = self._cursor_per_partition.get(self._to_partition_key(partition))
+            if not cursor:
+                cursor = self._create_cursor(self._NO_CURSOR_STATE)
+                self._cursor_per_partition[self._to_partition_key(partition)] = cursor
+
+            for cursor_slice in cursor.stream_slices(sync_mode, self._get_state_for_partition(partition)):
+                yield PerPartitionStreamSlice(partition, cursor_slice)
+
+    def update_cursor(self, stream_slice: PerPartitionStreamSlice, last_record: Optional[Record] = None):
+        if not last_record:
+            # The `update_cursor` method is called without `last_record` in order to set the initial state. In that case, stream_slice is
```

Comment:
should this union type be reflected in the method's signature?

Reply:
Isn't the Optional reflecting that? In any case, I would like to clean this in a follow-up PR and remove the fact that `last_record` is optional

Reply:
wrong line. I was referring to
>  In that case, stream_slice is not a PerPartitionStreamSlice but is a dict representing the state